File: 백준/Gold/2023. 신기한 소수/신기한 소수.py
import sys
input=sys.stdin.readline

#에라토스테네스의 체로 소수인지 확인
def is_prime(n):
  if n < 2:
      return False
  for i in range(2, int(n**0.5)+1):
      if n % i == 0:
          return False
  return True

n=int(input())

def sol(num):
  if len(str(num)) == n:
    print(num)
    return
  for i in range(10):
    k = num*10+i
    #num에 i 붙인 것이 소수일 경우
    if is_prime(k):
      sol(k)

#일의 자리 숫자 중 소수로 시
sol(2)
sol(3)
sol(5)
sol(7)

File: 백준/Gold/2023. 신기한 소수/신기한 소수.java
import java.io.*;
import java.util.*;

public class Main {
  static int N;
  public static void main(String args[]) throws IOException {
      BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
      
      N = Integer.parseInt(br.readLine());
      //소수 2, 3, 5, 7로 탐색 시작
      dfs(1, 2);
      dfs(1, 3);
      dfs(1, 5);
      dfs(1, 7);
      
      br.close();
  }
  
  static void dfs(int depth, int number){
      if(depth == N){
          if(isPrime(number)){
              System.out.println(number);
          }
          return;
      }
      for(int i = 1;i<10;i++){
          //짝수일 경우 탐색할 필요 없음
          if(i%2 == 0){
              continue;
          }
          if(isPrime(number*10+i)){
              dfs(depth+1,number*10+i);
          }
      }
  }
  
  static boolean isPrime(int num){
      for(int i = 2;i<=num/2;i++){
          if(num%i==0) return false;
      }
      return true;
  }
}